dist_sender : check rudimentary validation of batch

il9ue opened the pull request:
Do basic validation of the batch in `sendChunk`. Please give me any feedback, thanks!

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4529)

<!-- Reviewable:end -->

tschottdorf commented:
Thanks @il9ue! Could you move the verification into `keys.Range` (which would then return an error, dealing with a TODO on that function)? The upshot is that more places in the code would do this verification, and that we wouldn't iterate over the batch twice (`keys.Range` does it anyway).

il9ue commented:
Thanks for the comment @tschottdorf. I changed the place of verification inside ```key.Range ```. 
The logic verifies the batch right after ``` for loop ``` inside the ``` key.Range ``` , while if something goes wrong returns ``` emptyRSpan ```, if not returns actual range. Please give me feedbacks, Thanks!

tschottdorf commented on keys/keys.go:
```diff
@@ -442,9 +442,10 @@ func MakeSplitKey(key roachpb.Key) (roachpb.Key, error) {
 	for _, arg := range ba.Requests {
 		req := arg.GetInner()
-		if req.Method() == roachpb.Noop {
+		if req.Method() == roachpb.Noop && !roachpb.IsRange(req) {
```
what's this about?

il9ue replied:
This was test code, and i will omit later, sorry

tschottdorf commented on keys/keys.go:
```diff
@@ -462,5 +463,26 @@ func Range(ba roachpb.BatchRequest) roachpb.RSpan {
 	}
+
+	verifyRange := func(ba roachpb.BatchRequest) error {
+		for _, union := range ba.Requests {
```
please remove this loop; you're already iterating over the requests above. The verification can happen there, inline.
Also, please return an error from `keys.Range`.

il9ue replied:
ok, I will try to add using ```return error``` and update. Thanks

Commit:
dist_sender : check rudimentary validation of batch

il9ue commented:
@tschottdorf I have changed ``` keys.Range ``` logic and add ``` return error ```, also updated related files which call the function. Please check, thanks. (I accidently closed this PR and reopened, was mistake.)

tschottdorf commented on keys/keys.go:
```diff
@@ -471,16 +471,27 @@ func MakeSplitKey(key roachpb.Key) (roachpb.Key, error) {
 		}
 		h := req.Header()
 		key := Addr(h.Key)
+		if i == len(ba.Requests) {
```
this condition is never true, and we also want to check each request, so just remove line 484 and 494. You're also not using `key`, so move that definition below your check.

tschottdorf commented on keys/keys.go:
```diff
@@ -471,16 +471,27 @@ func MakeSplitKey(key roachpb.Key) (roachpb.Key, error) {
 		key := Addr(h.Key)
+		if i == len(ba.Requests) {
+			if bytes.Compare(h.Key, from) >= 0 {
+				return roachpb.RSpan{Key: from, EndKey: to}, util.Errorf("%s: must be less than KeyMax", h.Key)
```
on error, usually you return the zero values for all of the other return values, so `return nil, util.Errorf(...)`.

bdarnell commented on kv/dist_sender.go:
```diff
@@ -518,7 +518,10 @@ func (ds *DistSender) sendChunk(ctx context.Context, ba roachpb.BatchRequest) (*
 	// TODO(tschottdorf): consider rudimentary validation of the batch here
 	// (for example, non-range requests with EndKey, or empty key ranges).
-	rs := keys.Range(ba)
+	rs, rErr := keys.Range(ba)
```
All of these variables should be named `err`. We sometimes use other names when it's a different type (e.g. `pErr` for `roachpb.Error`) or in special cases, but most of the time a variable of type `error` should be named `err`.

il9ue commented:
@tschottdorf @bdarnell Thanks for the feedback. I have changed and updated few things in PR as requested. Please give a look!

tschottdorf commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[keys/keys.go, line 485 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529#-KBXB9nIxA22xP7ss_he:-KBXB9nIxA22xP7ss_hf:462391005) ([raw file](https://github.com/cockroachdb/cockroach/blob/b95a076382c852b9cd1dccb397edfc115e96a7be/keys/keys.go#L485)):*
remote `rsErr` and just return `nil`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529)*
<!-- Sent from Reviewable.io -->

il9ue commented:
Reviewed 4 of 6 files at r2, 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[keys/keys.go, line 484 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529#-KB3AgOZvobT32UbknXE-r2-484:-KBXu2148B8mILJZib3O:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/73c2f0d1eb24d270cb814638bd809e8372553c37/keys/keys.go#L484)):*
Done.

---

*[keys/keys.go, line 486 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529#-KB3AgOZvobT32UbknXE-r2-486:-KBXu2P9qDcbWerBR3jV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/73c2f0d1eb24d270cb814638bd809e8372553c37/keys/keys.go#L486)):*
Done.

---

*[keys/keys.go, line 485 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529#-KBXB9nIxA22xP7ss_he:-KBXtySOFAYRixKpfhYC:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b95a076382c852b9cd1dccb397edfc115e96a7be/keys/keys.go#L485)):*
Done.

---

*[kv/dist_sender.go, line 521 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529#-KBGmabPVlZLmKYDApz5-r2-521:-KBXu0YWEOv7SFgWVCBw:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/73c2f0d1eb24d270cb814638bd809e8372553c37/kv/dist_sender.go#L521)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[kv/dist_sender.go, line 521 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529#-KBGmabPVlZLmKYDApz5-r2-521:-KBYZIHvLSw86zdFCoYV:-515215468) ([raw file](https://github.com/cockroachdb/cockroach/blob/73c2f0d1eb24d270cb814638bd809e8372553c37/kv/dist_sender.go#L521)):*
There's still a lot of these left; all of the new `rErr` variables should be renamed to `err`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529)*
<!-- Sent from Reviewable.io -->

il9ue commented:
Review status: 2 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[kv/dist_sender.go, line 521 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529#-KBGmabPVlZLmKYDApz5-r2-521:-KB_xhhYYOIY18hSF_sM:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/73c2f0d1eb24d270cb814638bd809e8372553c37/kv/dist_sender.go#L521)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 2 files at r3, 1 of 1 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[keys/keys.go, line 490 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529#-KBiW_m1Rvjjz5XRVuv1:-KBiW_m1Rvjjz5XRVuv2:37845962) ([raw file](https://github.com/cockroachdb/cockroach/blob/494072d8f22e0108e88b5a75676fec02e89063a4/keys/keys.go#L490)):*
you just set `from = key`, so this isn't right. You need to compare with `roachpb.RKeyMax` every time (and it should be in the loop, not nested under `key.Less(from)`).

---

*[keys/keys.go, line 501 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529#-KBiWuvAV74P0gvcoRhr:-KBiWuvAV74P0gvcoRhs:-330895159) ([raw file](https://github.com/cockroachdb/cockroach/blob/494072d8f22e0108e88b5a75676fec02e89063a4/keys/keys.go#L501)):*
same here.

---

*[storage/store_test.go, line 103 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529#-KBiW6GdByQupU5BbvMY:-KBiW6GdByQupU5BbvMZ:-677932004) ([raw file](https://github.com/cockroachdb/cockroach/blob/494072d8f22e0108e88b5a75676fec02e89063a4/storage/store_test.go#L103)):*
`lookup`, not `Lookup`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4529)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@kkaneda, need to clean up my queue - would you be able to help (i.e. ping and maybe fix and land) with this? No worries if not - just reassign me in that case. Thanks!

kkaneda commented:
Ah just noticed this.... Yes, I'll take over!

kkaneda commented:
I don't know how to update other people's forked repo, so just created #6349 . The change is now fairly simple and I'm not totally sure it's worth committing though..

tschottdorf commented:
Superseded by #6349.
